### ISSUE TITLE
Add a method for constructing a `Waker` from a `Fn()`

### ISSUE THREAD

**Issue by JoJoJet** [rust-lang/libs-team#415]
# Proposal

## Problem statement

When writing code that deals with the low-level futures API, users often need to create their own `Waker`s with custom behavior, which has always been a laborious process. The `Wake` trait, based on `Arc`, allows users to create custom wakers in safe code, however defining a new trait and implementing the trait is still a considerable amount of boilerplate, and the ceremony for constructing the inner type and converting it to a waker can feel a bit arcane since it relies on a `From` impl which is difficult to discover in documentation.

## Motivating examples or use cases

Existing crates in the ecosystem often want a quick, zero-boilerplate way of constructing wakers, often for unit tests or examples. Some notable examlps are `event-listener`, `futures-lite`, and `flume`. All of these rely on the `waker-fn` crate to encapsulate the boilerplate of implementing `Wake` and constructing an `Arc` that gets converted into a waker.

## Solution sketch

Add a method to `Waker` that converts a closure into a waker.

```rust
// core::task
impl Waker {
    pub fn from_fn<F>(f: F) -> Waker
    where
        F: Fn() + Send + Sync + 'static;
}
```

This is similar to the existing methods `core::iter::from_fn` and `core::future::poll_fn` which are used to transform a closure into a specific kind of abstraction.

## Alternatives

Don't add this. The current methods for constructing wakers are not ideal but they are workable. However, I believe the solution presented here is still worth adding, as the method is straightforward and something that I think many users will just expect to exist -- some people may be surprised that `Waker::from_fn` does not exist already.

## Links and related work

The [`waker-fn`](https://crates.io/crates/waker-fn) crate has 47 million downloads on crates.io at the time of writing, which indicates a desire for this functionality.

## What happens now?

This issue contains an API change proposal (or ACP) and is part of the libs-api team [feature lifecycle]. Once this issue is filed, the libs-api team will review open proposals as capability becomes available. Current response times do not have a clear estimate, but may be up to several months.

[feature lifecycle]: https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html

## Possible responses

The libs team may respond in various different ways. First, the team will consider the *problem* (this doesn't require any concrete solution or alternatives to have been proposed):

- We think this problem seems worth solving, and the standard library might be the right place to solve it.
- We think that this probably doesn't belong in the standard library.

Second, if there's a concrete solution:

- We think this specific solution looks roughly right, approved, you or someone else should implement this. (Further review will still happen on the subsequent implementation PR.)
- We're not sure this is the right solution, and the alternatives or other materials don't give us enough information to be sure about that. Here are some questions we have that aren't answered, or rough ideas about alternatives we'd want to see discussed.

**Comment by Amanieu** [comment]
We (@rust-lang/libs-api) are happy with this, but we would like a second opinion from @rust-lang/wg-async.

**Comment by traviscross** [comment]
@JoJoJet: If perhaps you could sketch out the implementation of this, that would be helpful in answering questions I have about the exact intended semantics here.

In particular, this is proposed above to live in `core`, like `from_fn` and `poll_fn`, but the implementation in the `waker-fn` crate is:

```rust
pub fn waker_fn<F: Fn() + Send + Sync + 'static>(f: F) -> Waker {
    Waker::from(Arc::new(Helper(f)))
}
```

**Comment by JoJoJet** [comment]
That's a good point. I suppose the fact that `Waker` lives in `core` means we couldn't just move the method definition into `alloc`. What if we make it a free function instead?

```rust
// alloc::task
pub fn wake_fn<F>(f: F) -> Waker
where
    F: Fn() + Send + Sync + 'static;
```

**Comment by traviscross** [comment]
@rustbot labels -I-async-nominated

We talked about this in our async call today.

One item we discussed was whether this is worth it, given the existence of the `Wake` trait.  One could of course write:

```rust
let waker = {
    struct S;
    impl Wake for S {
        fn wake(self: Arc<Self>) {
            todo!();
        }
    }
    S
};
```

In the end, though, we all agreed that being able to write instead:

```rust
let waker = alloc::task::wake_fn(|| todo!());
```

...is convenient enough, and similar enough to other conveniences like `poll_fn` and `iter::from_fn`, that this is well motivated.

We discussed how this could be a good motivating use case for anonymous trait impls, e.g.:

```rust
let waker = impl Wake {
    fn wake(self: Arc<Self>) {
        todo!();
    }
};
```

...and how this would be a feasible language feature to pursue.  Even here, though, we still felt that `wake_fn` would add value.

Note that, as discussed in earlier comments in this thread, `wake_fn` with the semantics proposed has to live in `alloc` rather than in `core` (as `Waker` does) due to its use of `Arc`.

For code that must or wishes to only use `core`, we discussed how in addition to the `Arc` version, we could perhaps also see value in having this version in `core`:

```rust
// core::task
impl Waker {
    pub fn from_fn(f: fn()) -> Waker;
}
```

The provided function would of course either be stateless or would modify state in a `static`.  We were quickly able to come up with use cases for this.

*In summary:*

- WG-async supports `alloc::task::wake_fn`.
- WG-async is also separately open to `core::task::Waker::from_fn`.

**Comment by JoJoJet** [comment]
```rust
// core::task
impl Waker {
    pub fn from_fn(f: fn()) -> Waker;
}
```

This associated function makes a lot of sense to me, though the fact that it's much more discoverable than `wake_fn` might be a footgun. I could see someone easily getting confused if they try to call `Waker::from_fn` using a closure and don't realize that `wake_fn` exists -- it might be good to name it `fn from_static(f: fn())`

**Comment by traviscross** [comment]
That might be a benefit, actually, since it would be the obvious place to cross-link from the `Waker` page in the docs to `alloc::task::wake_fn`.  "If you want to pass in a closure, use *`wake_fn`* instead."

**Comment by Amanieu** [comment]
We discussed this in the libs-api meeting today. There were concerns about the utility of `Waker::from_fn` since it's a subset of `wake_fn` but there may be value in having a version that is guaranteed not to allocate. As such we're happy to accept both of these for now, with another review at stabilization time.
